File: client/app/actions/waitlist.ts
"use server";

import pool from "@/lib/db";

interface WaitlistResult {
  ok: boolean;
  error?: string;
}

export async function joinWaitlist(
  email: string,
  turnstileToken: string,
): Promise<WaitlistResult> {
  // ── Validate inputs ───────────────────────────────
  if (!email || !turnstileToken) {
    return { ok: false, error: "Missing required fields." };
  }

  // ── Verify Turnstile token ────────────────────────
  const secret = process.env.TURNSTILE_SECRET_KEY;
  if (!secret) {
    console.error("TURNSTILE_SECRET_KEY not set");
    return { ok: false, error: "Server misconfigured." };
  }

  const verification = await fetch(
    "https://challenges.cloudflare.com/turnstile/v0/siteverify",
    {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        secret,
        response: turnstileToken,
      }),
    },
  );

  const outcome = (await verification.json()) as { success: boolean };

  if (!outcome.success) {
    return { ok: false, error: "Bot verification failed. Please try again." };
  }

  // ── Insert into Postgres ──────────────────────────
  try {
    await pool.query(
      `INSERT INTO waitlist (email) VALUES ($1) ON CONFLICT (email) DO NOTHING`,
      [email.toLowerCase().trim()],
    );
    return { ok: true };
  } catch (err) {
    console.error("Waitlist insert error:", err);
    return { ok: false, error: "Something went wrong. Please try again." };
  }
}

File: backend/app/routers/waitlist.py
import httpx
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, EmailStr
from sqlalchemy.exc import IntegrityError

from ..config import settings
from ..db import get_session
from ..models import WaitlistEntry

router = APIRouter(prefix="/waitlist")
http = httpx.AsyncClient()


class WaitlistIn(BaseModel):
    email: EmailStr
    turnstile_token: str


async def verify_turnstile(token: str) -> None:
    if not settings.turnstile_secret_key:
        raise HTTPException(500, "Turnstile is not configured")

    response = await http.post(
        "https://challenges.cloudflare.com/turnstile/v0/siteverify",
        data={"secret": settings.turnstile_secret_key, "response": token},
    )

    if not response.json().get("success"):
        raise HTTPException(400, "Bot verification failed")


@router.post("")
async def join_waitlist(body: WaitlistIn) -> dict:
    await verify_turnstile(body.turnstile_token)

    db = get_session()
    try:
        db.add(WaitlistEntry(email=body.email.lower().strip()))
        db.commit()
    except IntegrityError:
        db.rollback()

    return {"ok": True}
